cpp/include/cudf/utilities/error.hpp

```diff
@@ -114,9 +114,8 @@ namespace detail {
 
 inline void throw_cuda_error(cudaError_t error, const char* file, unsigned int line)
 {
-  // Calls cudaGetLastError twice. It is nearly certain that a fatal error occurred if the second
+  // Calls cudaGetLastError again. It is nearly certain that a fatal error occurred if the second
   // call doesn't return with cudaSuccess.
-  cudaGetLastError();
   auto const last = cudaGetLastError();
```

Comment:
While the test case of fatal error can only be passed when I remove this line.
```c++
TEST(FatalCase, CudaFatalError)
{
  auto type = cudf::data_type{cudf::type_id::INT32};
  auto cv   = cudf::column_view(type, 256, (void*)256);
  cudf::binary_operation(cv, cv, cudf::binary_operator::ADD, type);
  EXPECT_THROW(CUDF_CUDA_TRY(cudaDeviceSynchronize()), cudf::fatal_cuda_error);
}
```

Reply:
Though you'd need to do this in a death test because it will corrupt the process context and leave the GPU unusable. See https://github.com/rapidsai/cudf/blob/1db83e3cdcc9fa74bda4c84e3a88bd8f329c4c2b/cpp/tests/error/error_handling_test.cu#L89-L111 for example.

Reply:
I tried with `fatal_kernel`, but it throwed nothing.

Reply:
>I tried with fatal_kernel, but it throwed nothing.

I'm still not comfortable with removing a line of code without understanding _why_ we're removing it.  It may have helped your test case, but we need to understand how that was significant for that test (is it a problem with the test?) or how removing this will not create problems in other scenarios trying to detect fatal errors.  If the CUDA error truly is fatal, it should not matter if we read the error an extra time.  It should make it even more likely it truly is a fatal error if the error persists despite extra attempts at clearing it.

Reply:
Hi @jlowe, according to the description in CUDA doc: `Returns the last error that has been produced by any of the runtime calls in the same host thread and resets it to cudaSuccess`,  the cudaGetLastError API works as poping the top error from the CUDA error stack ?

Reply:
> Hi @jlowe, according to the description in CUDA doc: Returns the last error that has been produced by any of the runtime calls in the same host thread and resets it to cudaSuccess, the cudaGetLastError API works like popping the top error from the CUDA error stack ?

Yes, normally it clears the error, but there are categories of errors that are _unclearable_.  These are the fatal errors we are trying to detect here.  If you're finding that `cudaGetLastError` is able to clear an error then it seems that error is not actually a fatal error and we should not report it as such.

Reply:
Hi @jrhemstad, thank you for the link. However, my simple tests suggested the second call of `cudaGetLastError` cleaned up fatal errors as well if I didn't misunderstand anything.
```c++
  // a valid CUDA call
  int* p0;
  EXPECT_EQ(cudaMalloc(&p0, 128), cudaSuccess);

  // produce an unrecoverable CUDA error: cudaErrorIllegalAddress
  auto type = cudf::data_type{cudf::type_id::INT32};
  auto cv   = cudf::column_view(type, 256, (void*)256);
  cudf::binary_operation(cv, cv, cudf::binary_operator::ADD, type);
  // wait the illegal binary operation to finish, then capture the CUDA status
  EXPECT_EQ(cudaDeviceSynchronize(), cudaErrorIllegalAddress);
  EXPECT_EQ(cudaGetLastError(), cudaErrorIllegalAddress);
  EXPECT_EQ(cudaGetLastError(), cudaSuccess); // the second call returns success

  // Any subsequent CUDA calls will fail, since the CUDA context has been corrupted.
  int* p1;
  EXPECT_EQ(cudaMalloc(&p1, 128), cudaErrorIllegalAddress);
  EXPECT_EQ(cudaGetLastError(), cudaErrorIllegalAddress);
  EXPECT_EQ(cudaGetLastError(), cudaSuccess); // the second call returns success

  int* p2;
  EXPECT_EQ(cudaMalloc(&p2, 128), cudaErrorIllegalAddress);
  EXPECT_EQ(cudaGetLastError(), cudaErrorIllegalAddress);
  EXPECT_EQ(cudaGetLastError(), cudaSuccess); // the second call returns success
```